packages/contracts/src/extensions/UniswapExtension.sol

```diff
@@ -53,19 +55,47 @@ contract UniswapExtension is Extension {
         require(!hasEmailRecipient, "recipient is not supported");
 
         (uint256 tokenInAmount, string memory tokenIn) = abi.decode(subjectParams[0], (uint256, string));
+        address tokenInAddr;
+        address tokenOutAddr;
+        // To avoid stack too deep, we use the following bracket
+        {
+            string memory tokenOut = abi.decode(subjectParams[1], (string));
+            tokenInAddr = tokenRegistry.getTokenAddress(tokenIn);
+            tokenOutAddr = tokenRegistry.getTokenAddress(tokenOut);
+            require(tokenOutAddr != address(0), "invalid out token name");
+        }
+
+        uint24 slippagePoints;
+        // To avoid stack too deep, we use the following bracket
+        {
+            if (templateIndex != 1 && templateIndex != 3) {
+                slippagePoints = slippageBasisPoints;
+            } else {
+                uint256 slippagePoints256 = abi.decode(subjectParams[2], (uint256));
+                // This value is user input * 10^18, we need to revert it as (user input * 10^2).
+                slippagePoints256 = slippagePoints256 / 10**16; 
+                require(slippagePoints256 <= type(uint24).max, "slippagePoints256 argument overflow detected");
+                slippagePoints = uint24(slippagePoints256);
+            }
+        }
 
         uint160 sqrtPriceLimitX96;
-        string memory tokenOut = abi.decode(subjectParams[1], (string));
-        if (templateIndex == 0) {
-            sqrtPriceLimitX96 = 0;
-        } else {
-            uint256 sqrtPriceLimitX96Uint256 = abi.decode(subjectParams[2], (uint256));
-            require(sqrtPriceLimitX96Uint256 <= type(uint160).max, "sqrtPriceLimitX96 argument overflow detected");
-            sqrtPriceLimitX96 = uint160(sqrtPriceLimitX96Uint256);
+        // To avoid stack too deep, we use the following bracket
+        {
+            if (templateIndex < 2) {
```

Comment:
I think its much easier to read the code if we split the entire execute() by if/else on teamplateId.

I refactored NFTExtension to be this way - https://github.com/zkemail/email-wallet/blob/audit-fix/packages/contracts/src/extensions/NFTExtension.sol

If there is like 1/2 lines of similar code across templates, I think its still ok. If there is more then this functionality could be extracted to a private function and called from each if/else.

Reply:
@saleel I agree with you, I've just update the one, thank you.